Clarify use of import via pseudo dir... (#18569)
nim-lang/Nim: doc/manual.rst

   its semantics are: *Use the search path to look for module name but ignore the standard
   library locations*. In other words, it is the opposite of `std`.

> It is recommended and preferred but not currently enforced that all imports include a "pseudo directorY" to qualify the search path in order to guaranty the correct search path to the desired module.

Reviewer: Well no, `std` is recommended, the other pseudo directories (you have a typo in "directorY" btw) are not recommended and hurt code evolution (code moves from Nimble package to somewhere else).

Author: I understood that there were only two "pseudo directories" `std/` and `pkg/` and an import should choose one or the other.  The manual says here that `std/` is the whole standard library and `pkg/` is everything except the standard library`.  What you say here seems to mean there are other "pseudo directories".  Do you mean core/pure/impure/js/wrappers/windows/posix, etc.?  I understood that these are just physical directories within the standard library code, and it would seem to be very bad form to use any of them, which I have never seen done.

It would seem to just be a coincidence that there is actually a `std/` real sub directory in the standard library and that there is also this `std/` "pseudo directory" as they aren't directly related to each other (other than not using **the `std/` pseudo directory doesn't search the real `std/` sub directory, which was my problem before**.

How about if, along with correcting the typo, I change the above to make it clear there is a choice between just the two "pseudo directories", either `std/` or `pkg/` as in the following:

"""
It is recommended and preferred but not currently enforced that all imports include either the `std/` or `pkg/`  "pseudo directory" as part of the import name to qualify the search path in order to guarantee the correct search path to the desired module.
"""

Reviewer 2: As far as I understand, `std/` should be used for all stdlib imports, but `pkg/` should only be used if a module name from a 3rd party package clashes with a local module name.

Author: Okay, it that is what is intended, I can do a further change little change to `doc/manual.rst`.  Can someone confirm?

Reviewer: That is what is intended indeed, I can confirm.

Reviewer 3: > but pkg/ should only be used if a module name from a 3rd party package clashes with a local module name.

this is a bad rule; it requires knowing about context (path, packages installed, file listing etc) to tell whether an import is local or from a 3rd party package (or from stdlib since std/ isn't always used).

Also, you're contradicting what you said here: https://github.com/nim-lang/RFCs/issues/267#issuecomment-825556683

Things work better when you don't need context to disambiguate, just like precedence rules work better without having to use a symbol table.

pkg/ should be recommended (and later, required with deprecation period) for external package imports

Reviewer: > Also, you're contradicting what you said here: nim-lang/RFCs#267 (comment)

What I said there is a proposal and potential solution for RFC 267. There is currently no guideline that says to prefer `pkg/`. We currently all agree on using `std`; we can recommend in the manual to use `pkg` *later*.

Reviewer 3: > we can recommend in the manual to use pkg later.

or we can do it now; the longer we wait, the more work there'll be to change imports to pkg. Adding pkg is forward and backward compatible so doesn't create complications.
We should recommend => deprecate => error

Reviewer: I cannot "recommend" what is unused by any codebase that I'm aware of.

Reviewer 4: Mandating `std/` I understand, but I think `pkg/` should only be used for disambiguation.

> It is recommended and preferred but not currently enforced that all module imports include either the std/ or pkg/ "pseudo directory" as part of the import name to qualify the search path in order to guarantee the correct search path to the desired module.

Reviewer: ```suggestion
It is recommended and preferred but not currently enforced that all stdlib module imports include the std/ "pseudo directory" as part of the import name.
```
